packages/core/source/index.ts

```diff
@@ -111,109 +102,131 @@ export default class Engine<Name extends string = string> {
+					utils.isAccessible(this.baseContext.parsedRules, '', name)
+			)
+		) as ParsedRules<Name>
 	}
 
 	resetCache() {
 		this.cache = emptyCache()
 	}
 
 	setSituation(
-		situation: Partial<Record<Name, PublicodesExpression | ASTNode>> = {}
+		situation: Partial<Record<Name, PublicodesExpression | ASTNode>> = {},
+		options: { keepPreviousSituation?: boolean } = {}
 	) {
 		this.resetCache()
-		this.parsedSituation = Object.fromEntries(
-			Object.entries(situation).map(([key, value]) => {
-				if (value && typeof value === 'object' && 'nodeKind' in value) {
-					return [key, value as ASTNode]
-				}
-				const parsedValue =
-					value && typeof value === 'object' && 'nodeKind' in value
-						? (value as ASTNode)
-						: this.parse(value, {
-								dottedName: `situation [${key}]`,
-								parsedRules: {},
-								...this.options,
-						  })
-				return [key, parsedValue]
-			})
-		)
-		return this
-	}
 
-	private parse(...args: Parameters<typeof parse>) {
-		return inlineReplacements(
-			this.replacements,
-			this.options.logger
-		)(
-			disambiguateReferenceAndCollectDependencies(
-				this.parsedRules,
-				{}
-			)(parse(...args))
+		const keepPreviousSituation = options.keepPreviousSituation ?? false
+
+		Object.keys(situation).forEach((name) => {
+			if (!(name in this.baseContext.parsedRules)) {
+				throw new Error(
+					`Erreur lors de la mise à jour de la situation : ${name} n'existe pas dans la base de règle.`
+				)
+			}
+			if (this.baseContext.parsedRules[name].private) {
+				throw new Error(
+					`Erreur lors de la mise à jour de la situation : ${name} est une règle privée (il n'est pas possible de modifier une règle privée).`
+				)
+			}
+		})
+
+		// The situation is implemented as a special sub namespace `$SITUATION`,
+		// present on each non-private rules
+		const situationToParse = Object.fromEntries(
+			Object.entries(situation).map(([nom, value]) => [
+				`[privé] ${nom} . $SITUATION`,
+				value && typeof value === 'object' && 'nodeKind' in value
+					? { valeur: value }
+					: value,
+			])
 		)
+
+		const savedBaseContext = copyContext(this.baseContext)
+
+		// const previousParsedRules = this.context.
+		this.context = {
+			...this.baseContext,
+			...parsePublicodes(
+				situationToParse as RawPublicodes<Name>,
+				keepPreviousSituation ? this.context : this.baseContext
+			),
+		}
+		this.baseContext = savedBaseContext
+
+		return this
 	}
 
 	inversionFail(): boolean {
 		return !!this.cache._meta.inversionFail
 	}
 
 	getRule(dottedName: Name): ParsedRules<Name>[Name] {
-		if (!(dottedName in this.parsedRules)) {
+		if (!(dottedName in this.baseContext.parsedRules)) {
 			throw new Error(`La règle '${dottedName}' n'existe pas`)
 		}
-		return this.parsedRules[dottedName]
-	}
 
-	getParsedRules(): ParsedRules<Name> {
-		return this.parsedRules
+		if (!(dottedName in this.publicParsedRules)) {
+			throw new Error(`La règle ${dottedName} est une règle privée.`)
```

Comment:
J'utilise `reduceAST` sur mesaidesvélo (pour le calcul des seuils de revenus) et je suis les références lors du parcours ce qui provoque l'erreur suivante :

> La règle aides . ile de france . $SITUATION est une règle privée

On a peut-être le cas sur mon-entreprise aussi ? Peut-être qu'on peut exposer un `reduceAST` qui évite d'aller dans des règles privées ?

(Évidemment pour mon cas précis je peux corriger de mon côté, surtout que c'est une API non documentée)

Reply:
Oui, il y a le même problème sur mon-entreprise... Je ne suis pas sûr exactement de quelle est la meilleure API pour récupérer les règles utilisée dans une autre. 

Je pense qu'on peut imaginer un `engine.getDependencies(rule)`. L'autre solution c'est de filtrer soit même les règles privée via `isAccessible`... 

Bref, c'est pas encore très clair, mais je partage ton besoin noté ici.